### PR TITLE
NFT & Fungible token contracts

### PR DESCRIPTION
**Pull request by fadeev**

### CODE REVIEW

**Review by andresaiello — APPROVED** [review]

File: contracts/nft/contracts/evm/UniversalNFT.sol [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        bytes calldata message
+    ) external payable onlyGateway returns (bytes4) {
+        if (context.sender != universal) revert Unauthorized();
+
+        (
+            address receiver,
+            uint256 tokenId,
+            string memory uri,
+            uint256 gasAmount,
+            address sender
```
[the remark this answers is not in the document]
Comment (fadeev) [reply contracts/nft/contracts/evm/UniversalNFT.sol:112]: c82fbd7

File: contracts/nft/contracts/example/Connected.sol [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+import "../evm/UniversalNFT.sol";
+
+contract Connected is UniversalNFT {
+    constructor(
+        address payable gatewayAddress,
+        address owner,
+        string memory name,
+        string memory symbol,
+        uint256 gasLimit
```
[the remark this answers is not in the document]
Comment (fadeev) [reply contracts/nft/contracts/example/Connected.sol:12]: 1953142

File: contracts/nft/contracts/zetachain/UniversalNFT.sol [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    ERC721Enumerable,
+    ERC721URIStorage,
+    Ownable2Step,
+    UniversalContract,
+    Events
+{
+    GatewayZEVM public immutable gateway;
+    address public immutable uniswapRouter;
+    uint256 private _nextTokenId;
+    bool public isUniversal = true;
```
[the remark this answers is not in the document]
Comment (fadeev) [reply contracts/nft/contracts/zetachain/UniversalNFT.sol]: a0c08da

File: contracts/nft/contracts/zetachain/UniversalNFT.sol [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+                    address(this),
+                    true,
+                    address(0),
+                    abi.encode(tokenId, uri, sender),
+                    0
+                )
+            );
+        }
+        emit TokenTransferToDestination(receiver, destination, tokenId, uri);
+    }
```
[the remark this answers is not in the document]
Comment (fadeev) [reply contracts/nft/contracts/zetachain/UniversalNFT.sol]: 8436b61

**Pull request merged**